### DISCUSSION TITLE
Auth API 소개 및 JWT 토큰 전략 (작성중)

### DISCUSSION THREAD

**Discussion by yongtaek2** [nhnacademy-be7-nuri-books/nuri-books-demo#42] [스터디]
### 1. 로그인 (JWT 발급)
- 인증된 Client에게 **JWT(Access Token & Refresh Token)를 발급**한다.
- **Access Token**을 사용하여 최초 로그인 이후 매 요청마다 인증을 처리한다.

### 2. 로그인 상태 연장 (JWT 재발급)
- **Refresh Token**을 활용하여 만료된 Access Token을 재발급받아 자동으로 갱신한다.

### 3. 로그아웃 (JWT 삭제)
- Client Side에 존재하는 Token을 삭제하고 Auth Server에서 사용하고있는 RDB/Redis에서도 해당 정보를 삭제한다.

### JWT 구조
- Header: 토큰 타입, 해싱 알고리즘으로 구성
- Payload: 토큰에 기입할 정보
  - 현재 Payload에서는 username, role, expiration 등을 관리하고 있음
  - 추후 username 대신 일회성 인조키를 사용할 것을 고민 중
- Signature: 헤더와 페이로드의 무결성을 검증하기위한 암호화 코드

### Access Token
- **인증 및 인가에 사용**하기 위한 토큰
- 발급 이후 유효시간은 **10~15분 정도로 짧게 설정**하여 탈취에 대한 대비를 한다.
- Authorization Header에 발급

### Refresh Token
- 만료된 **Access Token을 재발급**받기위한 보조 토큰
- 재발급의 용도로 사용되다보니 Access Token에 비해 긴 유효시간을 설정한다.
- Cookie에 발급
  - 내부적으로 Cookie관리가 비교적 복잡한 것으로 판단해 Header에 전송할 것을 고려 중

### Refresh Token with DB
- RDB 또는 Redis를 사용해 Refresh Token을 저장해야한다.
  - Server측에서 Token에 대한 제어권을 갖기 위함
- 어느 DB가 더 적절할지 고민 중

### Refresh Rotate
- 재발급된 Refresh Token은 유효기간이 남아있어도 새롭게 발급한다.
- 이를 통해 Refresh Token 자체에도 보안성을 높일 수 있다.

### Token Blacklist
- Access Token이 유효한 상태여도 특정 상황에 따라 해당 토큰을 차단하여 보안성을 높이는 방법
  - 로그아웃, 탈취, 계정 비활성화 등
- RDB 또는 Redis를 사용해 차단하려는 Access Token을 저장한다.
- 매 요청마다 Blacklist에서 조회하여 유효한 Access Token인지 이중 검사를 진행한다.

### Header에 저장하는 경우
- CSRF에 비교적 안전
- Header에 저장된 값을 JS로 꺼내 사용해야하므로 XSS에 상대적으로 취약

### Cookie에 저장하는 경우
- XSS에 비교적 안전
- 브라우저가 자동으로 요청에 포함시켜 CSRF에 상대적으로 취약

### 두 가지 경우 모두 사용
- 정해진 전략이 없어 때에 따라 적절한 방법을 선택할 수 있고 두 경우를 함께 사용하는 경우도 흔하게 볼 수 있다.
- 이 경우 공격 표적을 분산시킬 수 있다는 장점을 갖는다.

**Comment by rjsdn0124** [comment]
좋아요와 댓글!

**Reply by yongtaek2** [reply]
배우신 분!

**Comment by ljg0610** [comment]
깨달아버렸습니다...

**Reply by yongtaek2** [reply]
음성지원 완벽하고.
24시간 성장중.

**Comment by yeon-so** [comment]
짱! 잘보고 갑니다...

**Reply by yongtaek2** [reply]
인증 인가를 함께하는 게이트웨이 화이팅!
